pkg/storage/stores/shipper/bloomshipper/fetcher.go

```diff
@@ -125,120 +124,96 @@ func (f *Fetcher) writeBackMetas(ctx context.Context, metas []Meta) error {
 	return f.metasCache.Store(ctx, keys, data)
 }
 
-func (f *Fetcher) FetchBlocksWithQueue(ctx context.Context, refs []BlockRef) ([]BlockDirectory, error) {
-	responses := make(chan BlockDirectory, len(refs))
-	errors := make(chan error, len(refs))
-	for _, ref := range refs {
-		f.q.enqueue(downloadTask[BlockRef, BlockDirectory]{
+func (f *Fetcher) FetchBlocks(ctx context.Context, refs []BlockRef) ([]BlockDirectory, error) {
+	n := len(refs)
+
+	responses := make(chan downloadResponse[BlockDirectory], n)
+	errors := make(chan error, n)
+	for i := 0; i < n; i++ {
+		f.q.enqueue(downloadRequest[BlockRef, BlockDirectory]{
 			ctx:     ctx,
-			item:    ref,
-			key:     f.client.Block(ref).Addr(),
+			item:    refs[i],
+			key:     f.client.Block(refs[i]).Addr(),
+			idx:     i,
 			results: responses,
 			errors:  errors,
 		})
 	}
 
 	results := make([]BlockDirectory, len(refs))
-
-outer:
-	for i := 0; i < len(refs); i++ {
+	for i := 0; i < n; i++ {
 		select {
 		case err := <-errors:
 			return results, err
 		case res := <-responses:
-			for j, ref := range refs {
-				if res.BlockRef == ref {
-					results[j] = res
-					continue outer
-				}
-			}
-			return results, fmt.Errorf("no matching request found for response %s", res)
+			results[res.idx] = res.item
 		}
 	}
 
 	return results, nil
 }
 
-func (f *Fetcher) processTask(ctx context.Context, task downloadTask[BlockRef, BlockDirectory]) {
+func (f *Fetcher) processTask(ctx context.Context, task downloadRequest[BlockRef, BlockDirectory]) {
 	if ctx.Err() != nil {
 		task.errors <- ctx.Err()
 		return
 	}
 
-	refs := []BlockRef{task.item}
-	results, err := f.FetchBlocks(ctx, refs)
+	result, err := f.fetchBlock(ctx, task.item)
 	if err != nil {
 		task.errors <- err
 		return
 	}
 
-	for _, res := range results {
-		task.results <- res
+	task.results <- downloadResponse[BlockDirectory]{
+		item: result,
+		key:  task.key,
+		idx:  task.idx,
 	}
 }
 
-// FetchBlocks returns a list of block directories
-// It resolves them from three locations:
+// fetchBlock resolves a block from three locations:
 //  1. from cache
 //  2. from file system
 //  3. from remote storage
-func (f *Fetcher) FetchBlocks(ctx context.Context, refs []BlockRef) ([]BlockDirectory, error) {
-	if ctx.Err() != nil {
-		return nil, errors.Wrap(ctx.Err(), "fetch Blocks")
-	}
+func (f *Fetcher) fetchBlock(ctx context.Context, ref BlockRef) (BlockDirectory, error) {
```

Comment:
nit for later: we should probably implement these GETs over multiple blocks internally to allow us to use the `memcache.GetMulti` functionality to resolve multiple keys from a single request rather than 1:1